third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -172,6 +174,112 @@ pub fn serialized_size_allowing_delayed_values(
+            MoveStructLayout::RuntimeVariants(_) | MoveStructLayout::WithVariants(_),
+        )
+        | MoveTypeLayout::Vector(_) => 1,
+        // recursed:
+        MoveTypeLayout::Struct(MoveStructLayout::Runtime(fields)) => {
+            let mut total = 1; // Count the current visit, and aggregate all children
+            for field in fields {
+                total += type_visit_count_for_constant_serialized_size(field);
+            }
+            total
+        },
+        MoveTypeLayout::Struct(MoveStructLayout::WithFields(fields))
+        | MoveTypeLayout::Struct(MoveStructLayout::WithTypes { fields, .. }) => {
+            let mut total = 1; // Count the current visit, and aggregate all children
+            for field in fields {
+                total += type_visit_count_for_constant_serialized_size(&field.layout);
+            }
+            total
+        },
+        // Count the current visit, and inner visits
+        MoveTypeLayout::Native(_, inner) => {
+            1 + type_visit_count_for_constant_serialized_size(inner)
+        },
+    }
+}
+
+/// If given type has a constant serialized size (irrespective of the instance), it returns the serialized
+/// size in bytes any value would have.
+/// Otherwise it returns None.
+pub fn constant_serialized_size(ty_layout: &MoveTypeLayout) -> PartialVMResult<Option<usize>> {
+    let bcs_size_result = match ty_layout {
+        MoveTypeLayout::Bool => bcs::serialized_size(&false).map(Some),
+        MoveTypeLayout::U8 => bcs::serialized_size(&0u8).map(Some),
+        MoveTypeLayout::U16 => bcs::serialized_size(&0u16).map(Some),
+        MoveTypeLayout::U32 => bcs::serialized_size(&0u32).map(Some),
+        MoveTypeLayout::U64 => bcs::serialized_size(&0u64).map(Some),
+        MoveTypeLayout::U128 => bcs::serialized_size(&0u128).map(Some),
+        MoveTypeLayout::U256 => bcs::serialized_size(&u256::U256::zero()).map(Some),
+        MoveTypeLayout::Address => bcs::serialized_size(&AccountAddress::ZERO).map(Some),
+        // signer's size is VM implementation detail, and can change at will.
+        MoveTypeLayout::Signer => Ok(None),
+        // vectors have no constant size
+        MoveTypeLayout::Vector(_) => Ok(None),
+        // enums have no constant size
+        MoveTypeLayout::Struct(
+            MoveStructLayout::RuntimeVariants(_) | MoveStructLayout::WithVariants(_),
+        ) => Ok(None),
+        MoveTypeLayout::Struct(MoveStructLayout::Runtime(fields)) => {
+            let mut total = Some(0);
+            for field in fields {
+                let cur = constant_serialized_size(field)?;
+                match cur {
+                    Some(cur_value) => total = total.map(|v| v + cur_value),
+                    None => {
+                        total = None;
+                        break;
+                    },
+                }
+            }
+            Ok(total)
+        },
+        MoveTypeLayout::Struct(MoveStructLayout::WithFields(fields))
```

Comment:
WithFields and WithTypes should not occur at runtime, I think we can error. As `type_to_type_layout` returns the runtime variant